impl/src/test/java/io/kroxylicious/testing/kafka/KafkaClusterTest.java

```diff
@@ -112,6 +117,33 @@ public void kafkaClusterZookeeperRemoveBroker() throws Exception {
         }
     }
 
+    public static Stream<Arguments> restartBrokers() {
+        return Stream.of(
+                Arguments.of(1, true, true, (Predicate<Integer>) node -> true),
+                Arguments.of(2, true, true, (Predicate<Integer>) node -> node == 1),
+                Arguments.of(2, true, true, (Predicate<Integer>) node -> true),
+                Arguments.of(1, true, false, (Predicate<Integer>) node -> true),
+                Arguments.of(1, false, true, (Predicate<Integer>) node -> true));
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    public void restartBrokers(int brokersNum, boolean kraft, boolean abruptShutdown, Predicate<Integer> integerPredicate) throws Exception {
```

Comment:
The test doesn't actually verify that anything was restarted 😞 

The test would pass if `cluster.restartBrokers(integerPredicate, abruptShutdown);` was entirely a no-op.

Reply:
Totally agree with @SamBarker, we would need to check somehow that the brokers has been restarted or going down for a while. Is there any property that is changed when a broker is restarted (thinking about nodeId or something similar)?

Reply:
Thank you - that's a good spot.   It is not immediately obvious to me how to tell this from the kafka client.    AFAIK brokers don't have any kind of 'session/epoch' id available via the client.  Detecting the client's disconnection would be an imperfect proxy for a broker restart (but could suffer from false positives), but I don't actually see a clean way to do that either!

I'd rather not break the abstraction and start doing horrid things like searching for patterns in broker logs files.

Thinking hat time. Ideas welcome.

Reply:
I suppose one way to approach the problem might be to have the API accept a callback. 

```java
cluster.restartBrokers(Predicate<Integer>, boolean, Callable<Void> brokersDown);
```

 The semantics would be that the `restartBrokers` invokes `brokersDown` once the broker(s) are down. Restart wouldn't continue until the callback returns.  The test could then use the callback to make an independent verifications. 

It might have utility in Kroxylicious's test suite too - I can imagine it making an observation of say a target cluster's state, verifying that it has transitioned to, say,  'OFFLINE'.

Reply:
I like the suggestion I'm not entirely convinced about needing to wait for the callback to complete before continuing.

Reply:
I like it too. It is an smart way to say "Hey I'm down and going back online". Thanks @k-wall!

Reply:
What do you think to this solution? @SamBarker @franvila